### PR TITLE
CVE Scanning: Fix Vulnerabilities and Improve Local Utilities

### PR DESCRIPTION
**Pull request by sam-heilbron**
# Description

1. Fix vulnerabilities in 1.13 RC images
2. Improve utilities for scanning images locally

# Context
## Makefile improvements
Pulled in changes from https://github.com/solo-io/gloo/pull/7494

## Add Scanning Utility
We use a weekly job to identify vulnerabilities in images, but there is not an easy way for developers to scan Gloo images. 

I think having a CLI that can scan a batch of images for a particular version would be ideal to add to our scanning utility (https://github.com/solo-io/go-utils/tree/master/securityscanutils). In the meantime however, I decided to add it directly to projects.

I had previously added this to GlooE (https://github.com/solo-io/solo-projects/tree/master/hack/trivy) and we now introduce the same utility to Gloo.

### Testing
- [x] `make help` includes the new `scan-version` make target properly documented
- [x] `VERSION=1.13.0-rc1 make scan-version` runs scans against the Gloo images and outputs them to the `_output/scans` directory

## Fix Vulnerabilities
Remove the `x/mod` direct dependency and update some other dependencies to resolve the CVEs

### Testing
1. Scan 1.13.0-rc1 images
```
VERSION=1.13.0-rc1 make scan-version
```

```
PATH=/Users/samheilbron/go/src/github.com/solo-io/gloo/_output/.bin:$PATH GO111MODULE=on go run github.com/solo-io/go-utils/securityscanutils/cli scan-version -v \
                -r quay.io/solo-io\
                -t 1.13.0-rc1\
                --images gloo,gloo-envoy-wrapper,discovery,ingress,sds,certgen,access-logger,kubectl
{"level":"info","ts":"2022-12-05T13:37:06.137-0500","caller":"commands/scan_version.go:73","msg":"Starting ScanVersion with version=1.13.0-rc1"}
{"level":"debug","ts":"2022-12-05T13:37:07.340-0500","caller":"securityscanutils/trivy_scanner.go:77","msg":"Trivy found vulnerabilies after 1.202382497s in quay.io/solo-io/gloo:1.13.0-rc1"}
{"level":"debug","ts":"2022-12-05T13:37:07.340-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/gloo:1.13.0-rc1, ScanCompleted: true, VulnerabilityFound: true, Error: <nil>"}
{"level":"debug","ts":"2022-12-05T13:37:08.713-0500","caller":"securityscanutils/trivy_scanner.go:77","msg":"Trivy found vulnerabilies after 1.372851805s in quay.io/solo-io/gloo-envoy-wrapper:1.13.0-rc1"}
{"level":"debug","ts":"2022-12-05T13:37:08.713-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/gloo-envoy-wrapper:1.13.0-rc1, ScanCompleted: true, VulnerabilityFound: true, Error: <nil>"}
{"level":"debug","ts":"2022-12-05T13:37:09.975-0500","caller":"securityscanutils/trivy_scanner.go:77","msg":"Trivy found vulnerabilies after 1.262385305s in quay.io/solo-io/discovery:1.13.0-rc1"}
{"level":"debug","ts":"2022-12-05T13:37:09.975-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/discovery:1.13.0-rc1, ScanCompleted: true, VulnerabilityFound: true, Error: <nil>"}
{"level":"debug","ts":"2022-12-05T13:37:11.238-0500","caller":"securityscanutils/trivy_scanner.go:77","msg":"Trivy found vulnerabilies after 1.262496832s in quay.io/solo-io/ingress:1.13.0-rc1"}
{"level":"debug","ts":"2022-12-05T13:37:11.238-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/ingress:1.13.0-rc1, ScanCompleted: true, VulnerabilityFound: true, Error: <nil>"}
{"level":"debug","ts":"2022-12-05T13:37:12.494-0500","caller":"securityscanutils/trivy_scanner.go:77","msg":"Trivy found vulnerabilies after 1.255905841s in quay.io/solo-io/sds:1.13.0-rc1"}
{"level":"debug","ts":"2022-12-05T13:37:12.494-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/sds:1.13.0-rc1, ScanCompleted: true, VulnerabilityFound: true, Error: <nil>"}
{"level":"debug","ts":"2022-12-05T13:37:13.625-0500","caller":"securityscanutils/trivy_scanner.go:77","msg":"Trivy found vulnerabilies after 1.130851418s in quay.io/solo-io/certgen:1.13.0-rc1"}
{"level":"debug","ts":"2022-12-05T13:37:13.625-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/certgen:1.13.0-rc1, ScanCompleted: true, VulnerabilityFound: true, Error: <nil>"}
{"level":"debug","ts":"2022-12-05T13:37:14.899-0500","caller":"securityscanutils/trivy_scanner.go:77","msg":"Trivy found vulnerabilies after 1.274032312s in quay.io/solo-io/access-logger:1.13.0-rc1"}
{"level":"debug","ts":"2022-12-05T13:37:14.899-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/access-logger:1.13.0-rc1, ScanCompleted: true, VulnerabilityFound: true, Error: <nil>"}
{"level":"debug","ts":"2022-12-05T13:37:16.049-0500","caller":"securityscanutils/trivy_scanner.go:83","msg":"Trivy returned 0 after 1.150438335s on quay.io/solo-io/kubectl:1.13.0-rc1"}
{"level":"debug","ts":"2022-12-05T13:37:16.049-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/kubectl:1.13.0-rc1, ScanCompleted: true, VulnerabilityFound: false, Error: <nil>"}
{"level":"info","ts":"2022-12-05T13:37:16.050-0500","caller":"commands/scan_version.go:31","msg":"Vulernabilities found! Affected images: [quay.io/solo-io/gloo:1.13.0-rc1 quay.io/solo-io/gloo-envoy-wrapper:1.13.0-rc1 quay.io/solo-io/discovery:1.13.0-rc1 quay.io/solo-io/ingress:1.13.0-rc1 quay.io/solo-io/sds:1.13.0-rc1 quay.io/solo-io/certgen:1.13.0-rc1 quay.io/solo-io/access-logger:1.13.0-rc1]. Formatted results: _output/scans/1.13.0-rc1"}

```

2. Build images on this branch
```
VERSION=1.0.0-local make docker -B
```

3. Re-Run scan and confirm no vulnerabilities
```
VERSION=1.0.0-local make scan-version
```

```
PATH=/Users/samheilbron/go/src/github.com/solo-io/gloo/_output/.bin:$PATH GO111MODULE=on go run github.com/solo-io/go-utils/securityscanutils/cli scan-version -v \
                -r quay.io/solo-io\
                -t 1.0.0-local\
                --images gloo,gloo-envoy-wrapper,discovery,ingress,sds,certgen,access-logger,kubectl
{"level":"info","ts":"2022-12-05T13:51:41.579-0500","caller":"commands/scan_version.go:73","msg":"Starting ScanVersion with version=1.0.0-local"}
{"level":"debug","ts":"2022-12-05T13:51:44.875-0500","caller":"securityscanutils/trivy_scanner.go:83","msg":"Trivy returned 0 after 3.295580556s on quay.io/solo-io/gloo:1.0.0-local"}
{"level":"debug","ts":"2022-12-05T13:51:44.875-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/gloo:1.0.0-local, ScanCompleted: true, VulnerabilityFound: false, Error: <nil>"}
{"level":"debug","ts":"2022-12-05T13:51:47.203-0500","caller":"securityscanutils/trivy_scanner.go:83","msg":"Trivy returned 0 after 2.327388675s on quay.io/solo-io/gloo-envoy-wrapper:1.0.0-local"}
{"level":"debug","ts":"2022-12-05T13:51:47.203-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/gloo-envoy-wrapper:1.0.0-local, ScanCompleted: true, VulnerabilityFound: false, Error: <nil>"}
{"level":"debug","ts":"2022-12-05T13:51:49.376-0500","caller":"securityscanutils/trivy_scanner.go:83","msg":"Trivy returned 0 after 2.173018922s on quay.io/solo-io/discovery:1.0.0-local"}
{"level":"debug","ts":"2022-12-05T13:51:49.376-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/discovery:1.0.0-local, ScanCompleted: true, VulnerabilityFound: false, Error: <nil>"}
{"level":"debug","ts":"2022-12-05T13:51:51.419-0500","caller":"securityscanutils/trivy_scanner.go:83","msg":"Trivy returned 0 after 2.042993825s on quay.io/solo-io/ingress:1.0.0-local"}
{"level":"debug","ts":"2022-12-05T13:51:51.419-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/ingress:1.0.0-local, ScanCompleted: true, VulnerabilityFound: false, Error: <nil>"}
{"level":"debug","ts":"2022-12-05T13:51:53.262-0500","caller":"securityscanutils/trivy_scanner.go:83","msg":"Trivy returned 0 after 1.843159228s on quay.io/solo-io/sds:1.0.0-local"}
{"level":"debug","ts":"2022-12-05T13:51:53.262-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/sds:1.0.0-local, ScanCompleted: true, VulnerabilityFound: false, Error: <nil>"}
{"level":"debug","ts":"2022-12-05T13:51:55.140-0500","caller":"securityscanutils/trivy_scanner.go:83","msg":"Trivy returned 0 after 1.878025448s on quay.io/solo-io/certgen:1.0.0-local"}
{"level":"debug","ts":"2022-12-05T13:51:55.140-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/certgen:1.0.0-local, ScanCompleted: true, VulnerabilityFound: false, Error: <nil>"}
{"level":"debug","ts":"2022-12-05T13:51:57.045-0500","caller":"securityscanutils/trivy_scanner.go:83","msg":"Trivy returned 0 after 1.904483356s on quay.io/solo-io/access-logger:1.0.0-local"}
{"level":"debug","ts":"2022-12-05T13:51:57.045-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/access-logger:1.0.0-local, ScanCompleted: true, VulnerabilityFound: false, Error: <nil>"}
{"level":"debug","ts":"2022-12-05T13:51:57.394-0500","caller":"securityscanutils/trivy_scanner.go:83","msg":"Trivy returned 0 after 348.662515ms on quay.io/solo-io/kubectl:1.0.0-local"}
{"level":"debug","ts":"2022-12-05T13:51:57.394-0500","caller":"commands/scan_version.go:100","msg":"Scanned Image: quay.io/solo-io/kubectl:1.0.0-local, ScanCompleted: true, VulnerabilityFound: false, Error: <nil>"}
{"level":"info","ts":"2022-12-05T13:51:57.394-0500","caller":"commands/scan_version.go:31","msg":"No vulnerabilities found"}

```

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [ ] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -828,6 +828,10 @@ publish-security-scan:
 	gsutil cp -r $(SCAN_DIR)/gloo/markdown_results/** gs://$(SCAN_BUCKET)/gloo
 	gsutil cp -r $(SCAN_DIR)/solo-projects/markdown_results/** gs://$(SCAN_BUCKET)/solo-projects
 
+.PHONY: scan-version
```
Comment (nfuden) [inline Makefile:831]: nit: As we add new tools to our make file if they are not directly related to development defaults could we perhaps put them in a seperate go mod. 
In this case perhaps a make file for vulnerabilties or on-call?

File: hack/trivy/scan.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	"github.com/solo-io/go-utils/osutils/executils"
+	"github.com/solo-io/go-utils/securityscanutils"
+)
+
+const (
+	outputDir = "_output/scans"
+	imageRepo = "quay.io/solo-io"
+)
+
+// selected from a recent scan result: https://github.com/solo-io/gloo/issues/7361
```
Comment (bewebi) [inline hack/trivy/scan.go:22]: Are there other images in the repo that we don't scan?

File: hack/trivy/scan.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	}()
+
+	versionedOutputDir := path.Join(outputDir, version)
+	contextutils.LoggerFrom(ctx).Infof("Results will be written to %s", versionedOutputDir)
+	err = os.MkdirAll(versionedOutputDir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	var scanResults error
```
Comment (bewebi) [inline hack/trivy/scan.go:55]: I'm not sure that a vulnerability scanner constitutes an error per se...in fact it's intended behavior
When I run 
```VERSION=1.13.0-rc1 make scan-version```
I get
```
2022/11/30 19:04:49 unable to run: 7 errors occurred:
	* vulnerabilities found for quay.io/solo-io/gloo:1.13.0-rc1
	* vulnerabilities found for quay.io/solo-io/gloo-envoy-wrapper:1.13.0-rc1
	* vulnerabilities found for quay.io/solo-io/discovery:1.13.0-rc1
	* vulnerabilities found for quay.io/solo-io/ingress:1.13.0-rc1
	* vulnerabilities found for quay.io/solo-io/sds:1.13.0-rc1
	* vulnerabilities found for quay.io/solo-io/certgen:1.13.0-rc1
	* vulnerabilities found for quay.io/solo-io/access-logger:1.13.0-rc1


exit status 1
make: *** [scan-version] Error 1
```
which makes it seem like the scan itself failed, when really it did its job
Can we ensure that when the scan completes successfully, having identified CVEs, the output makes it clear that that is the case?

**Review by bewebi — CHANGES_REQUESTED** [review]
The tool appears to function as intended, however identifying a vulnerability should not be considered (or present as) an error with the tool

File: hack/trivy/scan.go (hunk 2 above)
Comment (sam-heilbron) [reply hack/trivy/scan.go:22]: Not to my knowledge, no. I used this issue and also checked the scan job logs to confirm

File: hack/trivy/scan.go (hunk 3 above)
Comment (sam-heilbron) [reply hack/trivy/scan.go:55]: Works for me!

File: Makefile (hunk 1 above)
Comment (sam-heilbron) [reply Makefile:831]: Yeah I like the idea. To be honest, I'm not sure what the ideal breakdown of separate makefiles would be, so i figured to keep it in one for now, close to the other security commands, and then we can migrate a whole bunch once we decide on the breakdown

**Review by bewebi — APPROVED** [review]
Looking good!

**Review by bewebi — DISMISSED** [review]

**Review by bewebi — APPROVED** [review]

**Review by nfuden — APPROVED** [review]

**Pull request merged**